include/hip/hcc_detail/program_state.hpp

```diff
@@ -93,16 +128,517 @@ class Kernel_descriptor {
     }
 };
 
-const std::unordered_map<hsa_agent_t, std::vector<hsa_executable_t>>& executables(
-    bool rebuild = false);
-const std::unordered_map<std::uintptr_t, std::vector<std::pair<hsa_agent_t, Kernel_descriptor>>>&
-functions(bool rebuild = false);
-const std::unordered_map<std::uintptr_t, std::string>& function_names(bool rebuild = false);
-std::unordered_map<std::string, void*>& globals(bool rebuild = false);
+template<typename P>
+inline
+ELFIO::section* find_section_if(ELFIO::elfio& reader, P p) {
+    const auto it = std::find_if(
+        reader.sections.begin(), reader.sections.end(), std::move(p));
+
+    return it != reader.sections.end() ? *it : nullptr;
+}
+
+inline
+__attribute__((visibility("hidden")))
```

Comment:
@AlexVlx could you say more about the rule for functions inside `hip_impl` namespace to carry `hidden` visibility or not? My understanding is that as long as they carry static data structures they are marked as hidden. Is this understanding correct?

Reply:
Yes, it is.